### PR TITLE
Docker upgrade on >= 16.04

### PR DESCRIPTION
**Pull request by lsh-0**
cc @giorgiosironi 

the bug mentions the problem disappears on service restart. If you're not ok with restarting docker on every highstate, we could find a condition that looks for a specific versions of the library, or only restart if the package gets updated, or ...

### CODE REVIEW

**Comment by giorgiosironi** [comment]
Wasn't clear on the desired behavior:
- 14.04: it's ok to keep the bug, we haven't practically encountered it and we are moving away from it
- 16.04: `elife-xpub` encountered the bug, it can upgrade Docker and it should upgrade Docker, but to a pinned version. I wasn't able to pin the version.
- 18.04: same as 16.04

**Comment by lsh-0** [comment]
take another look. I'm not pinning on 1604/1804 but I am now requiring the latest available. I don't think pinning `docker-ce` is the right solution here

**Comment by lsh-0** [comment]
er, looks like we have a wonderful new error:

```

 Error creating token: Error making API request.
 
 URL: POST https://master-server.elifesciences.org:8200/v1/auth/token/create
 Code: 403. Errors:
 
 * permission denied
 !!! Parallel execution exception under host '3.92.23.119':
 Process 3.92.23.119:
 Traceback (most recent call last):
   File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
     self.run()
   File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
     self._target(*self._args, **self._kwargs)
   File "/ext/srv/builder/venv/local/lib/python2.7/site-packages/fabric/tasks.py", line 241, in inner
     submit(task.run(*args, **kwargs))
   File "/ext/srv/builder/venv/local/lib/python2.7/site-packages/fabric/tasks.py", line 173, in run
     return self.wrapped(*args, **kwargs)
   File "/ext/srv/builder/venv/local/lib/python2.7/site-packages/fabric/decorators.py", line 185, in inner
     return func(*args, **kwargs)
   File "/ext/srv/builder/src/buildercore/core.py", line 314, in single_node_work
     return workfn(**work_kwargs)
   File "/ext/srv/builder/src/buildercore/bootstrap.py", line 517, in _update_ec2_node
     vault_token = vault.token_create(context['vault']['address'], vault.SALT_MASTERLESS_POLICY, display_name=context['stackname'])
   File "/ext/srv/builder/src/buildercore/vault.py", line 11, in token_create
     result = json.loads(external.execute(cmd))
   File "/ext/srv/builder/src/buildercore/external.py", line 4, in execute
     return subprocess.check_output(cmd)
   File "/usr/lib/python2.7/subprocess.py", line 574, in check_output
     raise CalledProcessError(retcode, cmd, output=output)
 CalledProcessError: Command '['vault', 'token', 'create', u'-address=https://master-server.elifesciences.org:8200', '-policy=master-server', u'-display-name=heavybox--builder-base-formula-pr-215-base-update', '-format=json']' returned non-zero exit status 2
 7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ==" >> /etc/ssh/ssh_known_hosts
 [3.92.23.119] out: + echo 'github.com,192.30.252.128 ssh-rsa AAAAB3NzaC1yc2EAAAABIwAAAQEAq2A7hRGmdnm9tUDbO9IDSwBK6TbQa+PXYPCPy6rbTrTtw7PHkccKrpp0yVhp5HdEIcKr6pLlVDBfOLX9QUsyCOV0wzfjIJNlGEYsdlLJizHhbn2mUjvSAHQqZETYP81eFzLQNnPHt4EVVUh7VfDESU84KezmD5QlWpXLmvU31/yMf+Se8xhHTvKSCZIFImWwoG6mbUoWf9nzpIoaSjB+weqqUUmpaaasXVal72J+UX2B+2RPW3RcT0eOzQgqlJL3RKrTJvdsjE3JEAvGq3lGHSZXy28G3skua2SmVi/w4yCE6gbODqnTWlg7+wC604ydGXA8VJiS5ap43JXiUFFAaQ=='
 [3.92.23.119] out: 
 [3.92.23.119] out: 
 
 [3.92.23.119] sudo: rm /tmp/bootstrap.sh-20190325005757
 [3.92.23.119] put: <file obj> -> /root/.ssh/id_rsa
 [3.92.23.119] sudo: rm -f /root/.ssh/id_rsa.pub && chown root:root /root/.ssh/id_rsa && chmod 600 /root/.ssh/id_rsa
 Traceback (most recent call last):
   File "/ext/srv/builder/venv/local/lib/python2.7/site-packages/fabric/main.py", line 757, in main
     *args, **kwargs
   File "/ext/srv/builder/venv/local/lib/python2.7/site-packages/fabric/tasks.py", line 426, in execute
     results['<local-only>'] = task.run(*args, **new_kwargs)
   File "/ext/srv/builder/venv/local/lib/python2.7/site-packages/fabric/tasks.py", line 173, in run
     return self.wrapped(*args, **kwargs)
   File "/ext/srv/builder/src/decorators.py", line 58, in wrap2
     return func(pname, *args, **kwargs)
   File "/ext/srv/builder/src/masterless.py", line 22, in wrapper
     return fn(*args, **kwargs)
   File "/ext/srv/builder/src/masterless.py", line 107, in launch
     bootstrap.update_stack(stackname, service_list=['ec2', 'sqs', 's3'], formula_revisions=formula_revisions)
   File "/ext/srv/builder/src/buildercore/bootstrap.py", line 398, in update_stack
     fn(stackname, context, **actual_arguments)
   File "/ext/srv/builder/src/buildercore/context_handler.py", line 84, in decorated_with_only_if
     return fn(stackname, context, **kwargs)
   File "/ext/srv/builder/src/buildercore/decorators.py", line 18, in _wrapper
     return func(stackname, *args, **kwargs)
   File "/ext/srv/builder/src/buildercore/bootstrap.py", line 552, in update_ec2_stack
     stack_all_ec2_nodes(stackname, _update_ec2_node, username=BOOTSTRAP_USER, concurrency=concurrency)
   File "/ext/srv/builder/src/buildercore/core.py", line 342, in stack_all_ec2_nodes
     return parallel_work(single_node_work, params)
   File "/ext/srv/builder/src/buildercore/core.py", line 355, in parallel_work
     return execute(parallel(single_node_work), hosts=list(params['public_ips'].values()))
   File "/ext/srv/builder/venv/local/lib/python2.7/site-packages/fabric/tasks.py", line 411, in execute
     ran_jobs = jobs.run()
   File "/ext/srv/builder/venv/local/lib/python2.7/site-packages/fabric/job_queue.py", line 167, in run
     self._fill_results(results)
   File "/ext/srv/builder/venv/local/lib/python2.7/site-packages/fabric/job_queue.py", line 190, in _fill_results
     datum = self._comms_queue.get_nowait()
   File "/usr/lib/python2.7/multiprocessing/queues.py", line 152, in get_nowait
     return self.get(False)
   File "/usr/lib/python2.7/multiprocessing/queues.py", line 135, in get
     res = self._recv()
 TypeError: ('__init__() takes at least 3 arguments (1 given)', <class 'subprocess.CalledProcessError'>, ())
[2019-03-25T00:59:26.661Z] Disconnecting from elife@184.73.15.149... done
```

checking state of vault ...

**Comment by lsh-0** [comment]
running the command manually on the master-server appears to get me a valid result. 

```bash
vault token create -address=https://master-server.elifesciences.org:8200 -policy=master-server -display-name=heavybox--builder-base-formula-pr-215-base-update -format=json
```

the rest of the output appears to just be stacktraces and noise. kicking the build

**Comment by lsh-0** [comment]
same problems. I thought the command was being called from the master-server rather than the instance being created because of a comment in the code.

can't ssh into heavybox PR instance, will attempt to re-create on an adhoc instance

**Comment by lsh-0** [comment]
my adhoc instance deploys perfectly. the output diverges here though:

```
[2019-03-25T01:30:15.502Z] INFO - buildercore.bootstrap - upload master builder key to /root/.ssh/id_rsa
[2019-03-25T01:30:15.502Z] INFO - buildercore.utils - uploading <_io.BytesIO object at 0x7f3fe0357470> to /root/.ssh/id_rsa
[2019-03-25T01:30:16.413Z] Error creating token: Error making API request.
```
investigating what that is ..

**Comment by lsh-0** [comment]
I can't figure it out.

Ran highstate on Alfred, thinking maybe it's own Vault token is busted. It completed with no problem, so I went through the states and found this:

```
----------
          ID: vault-approle-vault-wrapper-smoke-test
    Function: cmd.run
        Name: bash -c "source /etc/profile && /usr/local/bin/vault.sh token lookup > /dev/null"
      Result: True
     Comment: Command "bash -c "source /etc/profile && /usr/local/bin/vault.sh token lookup > /dev/null"" run
     Started: 02:48:17.616440
    Duration: 325.949 ms
     Changes:   
              ----------
              pid:
                  3193
              retcode:
                  0
              stderr:
                  Error looking up token: Error making API request.
                  
                  URL: GET https://master-server.elifesciences.org:8200/v1/auth/token/lookup-self
                  Code: 403. Errors:
                  
                  * permission denied
              stdout:
```

rest of the states look find. Investigating this state and wrapper ..

**Comment by lsh-0** [comment]
The script looks like it's doing a lazy-login. The output on stderr is from running a command that will fail if not logged in, whereupon it logs itself in and proceeds to run the vault command.

I don't know what the problem is, over to you @giorgiosironi

**Comment by giorgiosironi** [comment]
For reference: rollback of https://github.com/elifesciences/issues/issues/4755#issuecomment-475591849 didn't work.

File: elife/docker.sls [hunk 1]
```diff
@@ -40,20 +42,26 @@ docker-gpg-key:
...
 
 docker-packages:
+    {% if osrelease == '14.04' %}
     pkg.installed:
-        - pkgs: 
+        - pkgs:
+            # bug1
             # poorly-patched vulnerability 'fix' breaks 3.3 kernels, like the one in the Ubuntu Trusty 14.04 LTS
-            {% if salt['grains.get']('oscodename') == 'trusty' %}
             - docker-ce: 18.06.1~ce~3-0~ubuntu
```
Comment (giorgiosironi) [inline elife/docker.sls]:
```
docker-ce:
[2019-03-25T11:09:30.381Z] [3.85.81.24] out:                   ----------
[2019-03-25T11:09:30.381Z] [3.85.81.24] out:                   new:
[2019-03-25T11:09:30.381Z] [3.85.81.24] out:                       18.06.1~ce~3-0~ubuntu
```

File: elife/docker.sls [hunk 2]
```diff
@@ -40,20 +42,26 @@ docker-gpg-key:
...
-            {% else %}
-            - docker-ce
-            {% endif %}
+    {% else %}
+    # bug2
+    # we need a version greater than '18.09.3' but can't specify that with a wildcard (*).
+    # https://github.com/moby/moby/issues/38249#issuecomment-474795342
+    pkg.latest:
+        - name: docker-ce
```
Comment (giorgiosironi) [inline elife/docker.sls]: 16.04:
```
              docker-ce:
[2019-03-25T11:05:43.220Z] [18.208.171.138] out:                   ----------
[2019-03-25T11:05:43.220Z] [18.208.171.138] out:                   new:
[2019-03-25T11:05:43.220Z] [18.208.171.138] out:                       5:18.09.3~3-0~ubuntu-xenial
```
`18.04`:
```
[2019-03-25T11:06:51.354Z] [3.95.66.172] out:               docker-ce:
[2019-03-25T11:06:51.354Z] [3.95.66.172] out:                   ----------
[2019-03-25T11:06:51.354Z] [3.95.66.172] out:                   new:
[2019-03-25T11:06:51.354Z] [3.95.66.172] out:                       5:18.09.3~3-0~ubuntu-bionic
```

Comment (lsh-0) [reply elife/docker.sls]: :+1:

**Pull request merged**